include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -985,40 +999,70 @@ struct __parallel_find_backward_tag
 #else
     using _AtomicType = oneapi::dpl::__internal::__difference_t<_RangeType>;
 #endif
-    using _Compare = oneapi::dpl::__internal::__pstl_greater;
 
     template <typename _DiffType>
     constexpr static _AtomicType __init_value(_DiffType)
     {
         return _AtomicType{-1};
     }
+
+    template <sycl::access::address_space _Space>
+    static void
+    __save_state_to_atomic(__dpl_sycl::__atomic_ref<_AtomicType, _Space>& __atomic, _AtomicType __new_state)
+    {
+        // As far as we make search from end to the begin of data, we should save the last (maximal) found state.
+        __atomic.fetch_max(__new_state);
+    }
+
+    template <typename _DiffType>
+    static auto
+    __eval_res(_AtomicType __result, _DiffType __rng_n)
+    {
+        return __result != __init_value(__rng_n) ? __result : __rng_n;
+    }
 };
 
 // Tag for __parallel_find_or for or-semantic
 struct __parallel_or_tag
 {
-    class __atomic_compare
-    {
-      public:
-        template <typename _LocalAtomic, typename _GlobalAtomic>
-        bool
-        operator()(const _LocalAtomic& __found_local, const _GlobalAtomic& __found) const
-        {
-            return __found_local == 1 && __found == 0;
-        }
-    };
-
     using _AtomicType = int32_t;
-    using _Compare = __atomic_compare;
 
     // The template parameter is intended to unify __init_value in tags.
     template <typename _DiffType>
     constexpr static _AtomicType __init_value(_DiffType)
     {
         return 0;
     }
+
+    template <sycl::access::address_space _Space>
+    static void
+    __save_state_to_atomic(__dpl_sycl::__atomic_ref<_AtomicType, _Space>& __atomic, _AtomicType /*__new_state*/)
+    {
+        // As far as we make search from end to the begin of data, we simple save the fact of finding.
```

Comment:
The processing order isn't relevant here. We might want to simplify this to something like: "Store that a match was found. Its position is not relevant for or semantics."

Reply:
Fixed